# Imports

In [6]:
# Import data analysis and visualization libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [7]:
# Imports for machine learning
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression

# Dataset Info

In [8]:
# Information About the Iris Dataset
iris_data = load_iris()
print(iris_data.DESCR)

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
                
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
    ============== ==== ==== ======= ===== ===========

In [10]:
# Instantiate Dataframes for the dataset amd target vector
X = pd.DataFrame(data=iris_data.data, columns=iris_data.feature_names) 
y = pd.DataFrame(data=iris_data.target, columns=["iris_type"])

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

In [12]:
# Viewing information about dataset and target vector
X.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

In [11]:
y.head()

iris_type
0          0
1          0
2          0
3          0
4          0

# Making the Model

In [14]:
# Randomly partition the dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)

In [15]:
# Restructure the y data so we can fit and predict the model on it 
y_train, y_test = np.ravel(y_train), np.ravel(y_test)

In [16]:
# Instantiation of the model, supplinh args for the solver and multi_class
logreg = LogisticRegression(random_state=0, solver="lbfgs", multi_class="multinomial")

In [19]:
# standardizing the data
from sklearn.preprocessing import scale
X_train_scaled = scale(X_train)
# I will NOT use the scaled x training data though, 
# because this step by itself throws off accuracy of the predictions

In [24]:
# Fitting the model on training data
logreg.fit(X_train, y_train)

/Users/zainraza/Downloads/dev/courses/DS/classification_tutorial/env/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='multinomial', n_jobs=None, penalty='l2',
                   random_state=0, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

# Making Predictions, Checking Results

In [25]:
# Make predictions on the testing data, because the model hasn't seen it before
y_pred = logreg.predict(X_test)

In [28]:
# Check 1: Look for matching values in the predictions and the test y values
y_pred

array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 2, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0, 2, 1, 0, 2, 2, 1, 0, 2])

In [29]:
y_test

array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 2, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0, 2, 1, 0, 2, 2, 1, 0, 1])

In [30]:
# Check 2: Check 1 by using .shape property
y_pred.shape

(38,)

In [31]:
y_test.shape

(38,)

In [33]:
# We can also grab the relative classification probabilities in an array, 
# rather than the assigned classes themselves.
logreg.predict_proba(X_test)

array([[1.17922388e-04, 5.61467204e-02, 9.43735357e-01],
       [1.26292326e-02, 9.60453076e-01, 2.69176916e-02],
       [9.84397411e-01, 1.56025501e-02, 3.85521957e-08],
       [1.25169336e-06, 2.31508254e-02, 9.76847923e-01],
       [9.70234900e-01, 2.97649377e-02, 1.62569141e-07],
       [2.01665199e-06, 5.94446983e-03, 9.94053514e-01],
       [9.81899507e-01, 1.81004229e-02, 7.04287691e-08],
       [2.84243189e-03, 7.47093540e-01, 2.50064028e-01],
       [1.50916144e-03, 7.38518909e-01, 2.59971930e-01],
       [2.05290442e-02, 9.35891937e-01, 4.35790187e-02],
       [9.22384402e-05, 1.59464222e-01, 8.40443540e-01],
       [6.98633156e-03, 8.09996346e-01, 1.83017322e-01],
       [4.08224311e-03, 7.93600620e-01, 2.02317137e-01],
       [3.05683925e-03, 7.60908560e-01, 2.36034601e-01],
       [3.87703025e-03, 7.10277309e-01, 2.85845661e-01],
       [9.82815578e-01, 1.71843652e-02, 5.65333914e-08],
       [6.72908291e-03, 7.56467769e-01, 2.36803149e-01],
       [1.14293523e-02, 8.45109

# Determing Accuracy

In [34]:
# Send test data to model using testing data
logreg.score(X_test, y_test)

# note: you must send the testing data for x, rather than the predicted y values

0.9736842105263158

## Is this a good score?

Generally, we are looking to beat a baseline score for the probability of accuracy.
In this case, that score is technicall 33.33%, because we are classifying between 3 classes. So all in all 97% is pretty good!